Revert gemfile dev priority

durandom opened the pull request:
I'd like to revert https://github.com/ManageIQ/manageiq/pull/13305

The reason was 
> Include Gemfile.dev early so that it can override gem dependencies such as manageiq-ui-classic.

But this is only true for dependencies that are wrapped in a `unless dependencies.detect` block. Other deps cannot be overridden anymore. Before it was possible though.

Before you could do
```ruby
dependencies.reject!{|d| d.name == "ansible_tower_client"}
gem "ansible_tower_client", :path => '../ansible_tower_client'
```

which is also the approach I take for my local plugin development.

The `unless` block in the Gemfile is for including the miq core Gemfile into a plugins [Gemfile](https://github.com/ManageIQ/manageiq-content/blob/master/Gemfile#L8).
Without this, it would fail.

@martinpovolny @Fryguy @bdunne

durandom commented:
@miq-bot add_labels developer, enhancement, euwe/no
@miq-bot assign @bdunne

durandom commented:
jfyi, I have this in my Gemfile.dev.rb and all vmdb plugins checked out at `path_to_manageiq/plugins`

```ruby
%w(ui-classic providers-amazon providers-azure providers-lenovo providers-vmware).each do |plugin|
 if s = dependencies.find{|d| d.name == "manageiq-#{plugin}"}
   unless s.source.is_a? Bundler::Source::Gemspec
     # only do this if we are running from manageiq
     dependencies.reject!{|d| d.name == "manageiq-#{plugin}"}
     gem "manageiq-#{plugin}", :path => File.expand_path("plugins/manageiq-#{plugin}", __dir__)
     puts "#{plugin} changed to plugins/manageiq-#{plugin}"
   else
     puts "#{plugin} leaving at #{s.source}"
   end
 end
end
```

bdunne commented on Gemfile:
```diff
@@ -38,7 +23,6 @@ unless dependencies.detect { |d| d.name == "manageiq-providers-azure" }
 end
 
 unless dependencies.detect { |d| d.name == "manageiq-ui-classic" }
-  #gem "manageiq-ui-classic", :path => "../manageiq-ui-classic/"
```
This should go :+1:

Fryguy replied:
LOL...I had that in a change locally that I thought I already pushed up but apparently never did... I was so confused.

durandom commented:
ok, I'll look into making it into a method

Fryguy commented:
I'm thinking I'd prefer a solution where we can override *any* gem.  I know for linux_admin development, for example, we have change the Gemfile as well, and neither the old nor new Gemfile.dev.rb locations permit that.  I kind of like the direction @bdunne is going, but perhaps it can be more generic.

So, I'm thinking leave it at the end like it was previously, but have a method `override_gem` that is defined at the end of the Gemfile...something like

Gemfile:

```ruby
# ... Beginning of file

def override_gem(name, *args)
  return unless (s = dependencies.detect { |d| d.name == name })

  unless s.source.is_a? Bundler::Source::Gemspec
    # only do this if we are running from manageiq
    dependencies.delete(s)
    gem name, *args
  end
end

dev_gemfile = File.expand_path("Gemfile.dev.rb", __dir__)
eval_gemfile(dev_gemfile) if File.exist?(dev_gemfile)

# Load other additional Gemfiles
Dir.glob("bundler.d/*.rb").each { |f| eval_gemfile(File.expand_path(f, __dir__)) }
```

Fryguy commented:
(I don't know if that `.delete` call works, but I was trimming it up a bit mostly for readability)

bdunne commented:
:+1: to @Fryguy comment

bdunne commented:
Also, I think this method should only be defined when `!ENV["CI"] && !Rails.env.production?`.  (I know you can't use the second one in the Gemfile, but you get the point.

Fryguy commented:
@bdunne Actually, maybe we *should* allow production.  I'm thinking about how in downstream we lock to specific versions and we "hack" the Gemfile.  It would be a lot nicer if we could use this override feature.  cc @simaishi Thoughts?

simaishi commented:
@Fryguy I like the idea of using override for downstream too.  We could add Gemfile.prod.rb in productization, in the same way dev.rb is done?  That's a lot nicer/clearner than using the hack we have!

Fryguy commented:
@simaishi We also have the bundler.d directory, so we could drop it in there as well.

durandom commented:
So, I just added the suggested `manageiq_plugin` method and saw the conversation already moved further... will update

durandom commented:
@bdunne @Fryguy pls check again.

> I don't know if that .delete call works, but I was trimming it up a bit mostly for readability.

It works, nice, 🙇 

> is_a? source has to be removed if we are trying to override real gems. I honestly don't know what that line does (probably needs a comment to explain)

I couldnt figure out either, why I had added it. IIRC it was about running `bundle` from core _and_ the plugin. But could not reproduce the error, so I removed it.

bdunne commented on Gemfile:
```diff
@@ -162,3 +146,30 @@ unless ENV["APPLIANCE"]
+# override_gem 'manageiq-ui-classic', :path => File.expand_path("../manageiq-ui-classic", __dir__))
+#
+def override_gem(name, *args)
+  return unless (dependency = dependencies.find { |d| d.name == name })
```
Don't like the `return nil` here...
I think it should `raise` when you try to override a gem that is not in the list. I'm assuming that this will silently move on.(maybe it was a typo)
Alternatively, it could just `gem name, *args` if not already listed.

Fryguy replied:
I like this idea.  When you are overriding, you are making a conscious decision to override and if the underlying assumption changes, you probably want to know about it.

durandom replied:
yep, I'm with the `raise` suggestion. Neat.

bdunne commented on Gemfile:
```diff
@@ -162,3 +146,30 @@ unless ENV["APPLIANCE"]
+#   To develop a gem locally and override its source to a checked out repo
+#   you can use this helper method in Gemfile.dev.rb e.g.
+#
+# override_gem 'manageiq-ui-classic', :path => File.expand_path("../manageiq-ui-classic", __dir__))
```
I think this comment and method should be [above the "Load developer specific Gemfile" comment](https://github.com/ManageIQ/manageiq/pull/13418/files#diff-8b7db4d5cc4b8f6dc8feb7030baa2478R153)

Fryguy commented:
@martinpovolny @dclarizio Once this gets merged, you'll probably have to change your `Gemfile.dev.rb` files to use the new helper method.  Will this be a problem?

cc @ManageIQ/core-providers You all will probably like this feature because you can point to a local path without modifying the Gemfile itself for cross-repo work.

martinpovolny commented:
@Fryguy : no problem, but we need to update the guides

durandom commented:
@bdunne updated with your`raise` suggestion

durandom commented:
https://github.com/ManageIQ/guides/pull/180 is the PR to guides

durandom commented:
yay, green :) @bdunne merge?

Fryguy commented:
LGTM... I want to merge this in a coordinated way so as not to mess up existing developers.  I am going to announce it and merge after that announcement

juliancheal commented:
![byeee](https://cloud.githubusercontent.com/assets/63991/21857092/32f7be08-d81c-11e6-88e7-4a679abdcdc6.gif)
@Fryguy like this? "Hey all, I'm merging this, Byeeee"

Fryguy commented:
@durandom @bdunne Do you think this could be marked euwe/yes?  I ask because when switching between branches, the Gemfile.dev.rb comes along for the ride (and thus will blow up on euwe).  I'm less concerned about darga, but perhaps there as well?

bdunne commented:
I'm fine with backporting to both.  The code will raise if the gem to be overridden isn't in the Gemfile.

durandom commented:
I agree with @bdunne   It's just a convenience method. 
But then you might don't want to override the ui for euwe because it's not extracted yet. 
So we might add some branch conditions to it :)

durandom commented:
> But then you might don't want to override the ui for euwe because it's not extracted yet.

So, I'm on `euwe` branch and have `override_gem 'manageiq-ui-classic', :path => ...` in my Gemfile.dev.rb, then bundle will exit with an error because we `raise` instead of `return`

Either we add something like 

```ruby
override_gem 'manageiq-ui-classic', :skip_branches => ['euwe', 'darga'], :path => '../'
```

which can become messy if you work on feature branches forked from euwe.

_OR_

we don't raise and return again.

I'm opting just for the returning to keep things simple. Maybe issue a warning (which will not be seen due to the amount of bundle output 😄 )

bdunne commented:
I still prefer the raise

Fryguy commented:
I could go either way, but I think the nice part of a raise is it forces the developer to deal with it.  They can then wrap their override_gem command in a `if File.read("VERSION").start_with?("euwe")` or whatever.  I think we can add the :skip_branches in a follow up pass if necessary.

Fryguy commented:
Darga backport details

```
commit 35fd0bd641b2ec12ee54eed92a601227e9ed6e74
Author: Jason Frey <fryguy9@gmail.com>
Date:   Thu Jan 12 11:33:51 2017 -0500

    Merge pull request #13418 from durandom/revert_gemfile_dev_priority

    Revert gemfile dev priority
    (cherry picked from commit f372770e06f18807c234b81b9f6411088a36f5d2)
```

Fryguy commented:
Euwe backport details

```
commit daa3cafc4b6683f72a5031009b5e657ffa572306
Author: Jason Frey <fryguy9@gmail.com>
Date:   Thu Jan 12 11:33:51 2017 -0500

    Merge pull request #13418 from durandom/revert_gemfile_dev_priority

    Revert gemfile dev priority
    (cherry picked from commit f372770e06f18807c234b81b9f6411088a36f5d2)
```
